From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Mon, 4 May 2026 22:37:32 +0800
Subject: [PATCH] Leaf Improve BlockEntity ticking isRemoved check

Uses direct reference for isRemoved check in BlockEntity ticking,
reducing cost caused by nested wrapper calls
Leaf (Before): ~15104ms
Leaf (After): ~628ms (-96%)
This should help for massive hopper chains or hopper matrix.

Co-authored by: kidofcubes <kidofcubes@gmail.com>
As part of: Leaf (https://github.com/Winds-Studio/Leaf/blob/ver/1.21.11/leaf-server/minecraft-patches/features/0209-Improve-BlockEntity-ticking-isRemoved-check.patch)
Licensed under: MIT (https://github.com/Winds-Studio/Leaf/blob/ver/1.21.11/licenses/MIT.txt)

diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index c44d6934984cbad249fe7e91c983678b62d8db2c..761dbf25e8bd7bfd7894da423f7797ebe1e059dd 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -1043,13 +1043,26 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
 
     private static class RebindableTickingBlockEntityWrapper implements TickingBlockEntity {
         private TickingBlockEntity ticker;
+        private @Nullable BlockEntity blockEntityReference = null; // Leaf - Improve BlockEntity ticking isRemoved check
 
         private RebindableTickingBlockEntityWrapper(final TickingBlockEntity ticker) {
             this.ticker = ticker;
+            // Leaf start - Improve BlockEntity ticking isRemoved check
+            if (ticker instanceof BoundTickingBlockEntity<?> boundTicker) {
+                this.blockEntityReference = boundTicker.blockEntity;
+            }
+            // Leaf end - Improve BlockEntity ticking isRemoved check
         }
 
         private void rebind(final TickingBlockEntity ticker) {
             this.ticker = ticker;
+            // Leaf start - Improve BlockEntity ticking isRemoved check
+            if (ticker instanceof BoundTickingBlockEntity<?> boundTicker) {
+                this.blockEntityReference = boundTicker.blockEntity;
+            } else {
+                this.blockEntityReference = null;
+            }
+            // Leaf end - Improve BlockEntity ticking isRemoved check
         }
 
         // Folia start - region threading
@@ -1066,6 +1079,11 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
 
         @Override
         public boolean isRemoved() {
+            // Leaf start - Improve BlockEntity ticking isRemoved check
+            if (this.blockEntityReference != null) {
+                return this.blockEntityReference.isRemoved();
+            }
+            // Leaf end - Improve BlockEntity ticking isRemoved check
             return this.ticker.isRemoved();
         }
 
